Source/CarthageKit/Project.swift

```diff
@@ -314,17 +304,53 @@ public final class Project { // swiftlint:disable:this type_body_length
 			}
 	}
 
+	/// Produces the sub dependencies of the given dependency. Uses the checked out directory if able
+	private func dependencySet(for dependency: Dependency, version: PinnedVersion) -> SignalProducer<Set<Dependency>, CarthageError> {
+		return self.dependencies(for: dependency, version: version, tryCheckoutDirectory: true)
+			.map { $0.0 }
+			.collect()
+			.map { Set($0) }
+			.concat(value: Set())
+			.take(first: 1)
+	}
+
 	/// Loads the dependencies for the given dependency, at the given version.
 	private func dependencies(for dependency: Dependency, version: PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError> {
+		return self.dependencies(for: dependency, version: version, tryCheckoutDirectory: false)
+	}
+
+	/// Loads the dependencies for the given dependency, at the given version. Optionally can attempt to read from the Checkout directory
+	private func dependencies(for dependency: Dependency, version: PinnedVersion, tryCheckoutDirectory: Bool) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError> {
 		switch dependency {
 		case .git, .gitHub:
 			let revision = version.commitish
-			return self.cloneOrFetchDependency(dependency, commitish: revision)
+			let cartfileFetch: SignalProducer<Cartfile, CarthageError> = self.cloneOrFetchDependency(dependency, commitish: revision)
 				.flatMap(.concat) { repositoryURL in
 					return contentsOfFileInRepository(repositoryURL, Constants.Project.cartfilePath, revision: revision)
 				}
 				.flatMapError { _ in .empty }
 				.attemptMap(Cartfile.from(string:))
+
+			let cartfileSource: SignalProducer<Cartfile, CarthageError>
+			if tryCheckoutDirectory {
```

Comment:
It's not taking the revision explicitly into account. However this is the same amount of verification that any step of the build phase takes. Right now we check the Cartfile.resolved for build order but build whatever happens to be sitting in  the checkout folder. So if we're assuming that we're building the right revision at that point, then it is no worse to also read the Cartfile from there.